cosmos/x/evm/keeper/abci.go

```diff
@@ -24,17 +24,36 @@ import (
 	"context"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/ethereum/go-ethereum/consensus/misc"
+
+	coretypes "pkg.berachain.dev/polaris/eth/core/types"
 )
 
 func (k *Keeper) BeginBlocker(ctx context.Context) error {
-	sCtx := sdk.UnwrapSDKContext(ctx)
+	if k.lock {
+		// unlock the keeper so that the Polaris services (EVM, JSON-RPC, TxPool, etc.) can start
+		k.lock = false
+	}
+
 	// Prepare the Polaris Ethereum block.
-	k.lock = false
-	k.polaris.Prepare(ctx, uint64(sCtx.BlockHeight()))
+	k.polaris.Prepare(ctx, uint64(sdk.UnwrapSDKContext(ctx).BlockHeight()))
 	return nil
 }
 
 func (k *Keeper) EndBlock(ctx context.Context) error {
 	// Finalize the Polaris Ethereum block.
-	return k.polaris.Finalize(ctx)
+	if err := k.polaris.Finalize(ctx); err != nil {
+		return err
+	}
+
+	// Prepare the txpool for the next pending block.
+	// TODO: move this.
+	chainConfig := k.host.GetConfigurationPlugin().ChainConfig()
+	k.GetHost().GetTxPoolPlugin().Prepare(
+		misc.CalcBaseFee(chainConfig, k.polaris.Blockchain().CurrentBlock()),
+		coretypes.LatestSignerForChainID(chainConfig.ChainID),
```

Comment:
use MakeSigner, you have the block number and time

Reply:
also why not just call this in chain writer

Reply:
dont wanna give blockchain access to the tx pool plugin in preparation for breaking it out properly